README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fauxnaif <img src="man/figures/logo.png?raw=TRUE" align="right" height="138" />

<!-- badges: start -->

[![](https://www.r-pkg.org/badges/version/fauxnaif?color=brightgreen)](https://cran.r-project.org/package=fauxnaif)
[![](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![License:
MIT](https://img.shields.io/badge/license-MIT-blueviolet.svg)](https://cran.r-project.org/web/licenses/MIT)
[![R build
status](https://github.com/rossellhayes/fauxnaif/workflows/R-CMD-check/badge.svg)](https://github.com/rossellhayes/fauxnaif/actions)
[![](https://app.codecov.io/gh/rossellhayes/fauxnaif/branch/main/graph/badge.svg)](https://app.codecov.io/gh/rossellhayes/fauxnaif)
[![Dependencies](https://tinyverse.netlify.com/badge/fauxnaif)](https://cran.r-project.org/package=fauxnaif)
<!-- badges: end -->

***faux-naïf***
([/ˌfoʊ.naɪˈif/](https://en.wikipedia.org/wiki/Help:IPA/English)): a
person who pretends to be simple or innocent

**fauxnaif**: an R package for simplifying data by pretending values are
`NA`

## Overview

**fauxnaif** provides an extension to `dplyr::na_if()`. Unlike
[**dplyr**](https://github.com/tidyverse/dplyr)’s `na_if()`,
`na_if_in()` allows you to specify multiple values to be replaced with
`NA` using a single function. **fauxnaif** also includes a complementary
function `na_if_not()` to specify values to keep.

## Installation

You can install `fauxnaif` from
[CRAN](https://cran.r-project.org/package=fauxnaif):

``` r
install.packages("fauxanif")
```

Or the development version from
[GitHub](https://github.com/rossellhayes/fauxnaif):

``` r
# install.packages("remotes")
remotes::install_github("rossellhayes/fauxnaif")
```

## Usage

``` r
library(dplyr)
library(fauxnaif)
```

### The basics

Let’s say we want to remove an unwanted negative value from a vector of
numbers

``` r
-1:10
#>  [1] -1  0  1  2  3  4  5  6  7  8  9 10
```

We can replace -1…

… explicitly:

``` r
na_if_in(-1:10, -1)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

… by specifying values to keep:

``` r
na_if_not(-1:10, 0:10)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

… using a formula:

``` r
na_if_in(-1:10, ~ . < 0)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

### A little more complex

``` r
messy_string <- c("abc", "", "def", "NA", "ghi", 42, "jkl", "NULL", "mno")
```

We can replace unwanted values…

… one at a time:

``` r
na_if_in(messy_string, "")
#> [1] "abc"  NA     "def"  "NA"   "ghi"  "42"   "jkl"  "NULL" "mno"
```

… or all at once:

``` r
na_if_in(messy_string, "", "NA", "NULL", 1:100)
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
na_if_in(messy_string, c("", "NA", "NULL", 1:100))
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
na_if_in(messy_string, list("", "NA", "NULL", 1:100))
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
```

… or using a clever formula:

``` r
grepl("[a-z]{3,}", messy_string)
#> [1]  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE
na_if_not(messy_string, ~ grepl("[a-z]{3,}", .))
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
```

### With data frames

``` r
faux_census
#> # A tibble: 5 × 4
#>   state    age  income gender                      
#>   <chr>  <dbl>   <dbl> <chr>                       
#> 1 TX        57 9999999 Gender is a social construct
#> 2 Canada    49  149000 Male                        
#> 3 NY       557   90750 f                           
#> 4 LA         2   61000 Male                        
#> 5 TN        64 9999999 M
```

na_if_in() is particularly useful inside `dplyr::mutate()`:

``` r
faux_census %>%
 mutate(
   income = na_if_in(income, 9999999),
   age    = na_if_in(age, ~ . < 18, ~ . > 120),
   state  = na_if_not(state, ~ grepl("^[A-Z]{2,}$", .)),
   gender = na_if_in(gender, ~ nchar(.) > 20)
 )
#> # A tibble: 5 × 4
#>   state   age income gender
#>   <chr> <dbl>  <dbl> <chr> 
#> 1 TX       57     NA <NA>  
#> 2 <NA>     49 149000 Male  
#> 3 NY       NA  90750 f     
#> 4 LA       NA  61000 Male  
#> 5 TN       64     NA M
```

Or you can use `dplyr::across()` on data frames:

``` r
faux_census %>%
  mutate(
    across(age, na_if_in, ~ . < 18, ~ . > 120),
    across(state, na_if_not, ~ grepl("^[A-Z]{2,}$", .)),
    across(where(is.character), na_if_in, ~ nchar(.) > 20),
    across(everything(), na_if_in, 9999999)
  )
#> # A tibble: 5 × 4
#>   state   age income gender
#>   <chr> <dbl>  <dbl> <chr> 
#> 1 TX       57     NA <NA>  
#> 2 <NA>     49 149000 Male  
#> 3 NY       NA  90750 f     
#> 4 LA       NA  61000 Male  
#> 5 TN       64     NA M
```

------------------------------------------------------------------------

Hex sticker fonts are [Bodoni\* by indestructible
type\*](https://github.com/indestructible-type/Bodoni) and [Source Code
Pro by Adobe](https://github.com/adobe-fonts/source-code-pro).

Image adapted from icon made by [Freepik](https://www.freepik.com) from
[flaticon.com](https://www.flaticon.com/free-icon/paper-shredder_1701401).

Please note that **fauxnaif** is released with a [Contributor Code of
Conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/).


NEWS.md
# fauxnaif 0.7.1

- Update Roxygen version to avoid CRAN NOTE.

# fauxnaif 0.7.0

-   `na_if_in()` and `na_if_not()` now support `vctrs` classes like `haven_labelled`.
-   Function arguments must now return logical vectors of the same length as `x`.
-   `fauxnaif` no longer throws warnings. Situations that previously produced a warning now produce an error:
    -   If no arguments are passed to `...`.
    -   If any arguments passed to `...` are not coercible to a vector or a function.
-   Removed deprecated `fauxnaif::na_if()` and scoped `na_if_*()` functions.
-   Added `pkgdown` URL to `DESCRIPTION`.
-   Added `inst/CITATION`.

# fauxnaif 0.6.1

-   Replace argument `input` with `x`.
-   Advance `fauxnaif::na_if()` and scoped `na_if_*()` functions from `deprecate_soft()` to `deprecate_warn()`.
-   Only warn `fauxnaif::na_if()` is deprecated if arguments could not be handled by `dplyr::na_if()`.
-   Added `pkgdown` site.
-   Remove `inst/CITATION`.

# fauxnaif 0.6.0

-   No longer produce a warning when no replacements are made.
-   Deprecate scoped na_if() functions and suggest replacing them with `dplyr::across()`.
-   Deprecate `fauxnaif::na_if()` and suggest replacing it with `na_if_in()`.

# fauxnaif 0.5.6

-   Initial CRAN release.
-   Scoped na_if\_\*() functions now call an internal function fauxnaif:::scoped_na_if() to reduce repeated code.

# fauxnaif 0.5.5

-   Refactored scoped na_if\_\*() to only message when no replacements are made in in entire data frame. Previously, a message would be generated for each column in which no replacements were made.

# fauxnaif 0.5.4

-   Changed all instances of `sapply()` to `vapply()` or `lapply()` for type stability.

# fauxnaif 0.5.3

-   Moved package imports from "Depends" to "Imports" for cleaner loading.

# fauxnaif 0.5.2

-   Included magrittr pipe.

# fauxnaif 0.5.1

-   Rewrote scoped na_if\_\* functions to use only exported dplyr functions.

# fauxnaif 0.5.0

-   Added support for function arguments.
-   Complete rewrite of underlying implementation function.

# fauxnaif 0.4.1

-   Added unit tests for scoped functions.

# fauxnaif 0.4.0

-   Added scoped variants of `na_if()` and `na_if_not()` based on scoped `dplyr::mutate_*()` functions.

# fauxnaif 0.3.0

-   Added a `NEWS.md` file to track changes to the package.


R/na_if.r
#' Convert values to NA
#'
#' This is a replacement for [dplyr::na_if()].
#' It is useful if you want to convert annoying values to `NA`.
#' Unlike [dplyr::na_if()], this function allows you to specify multiple values
#' to be replaced with `NA` at the same time.
#'  * `na_if_in()` replaces values that match its arguments with `NA`.
#'  * `na_if_not()` replaces values that *do not* match its arguments with `NA`.
#'
#' @section Formulas:
#' These functions accept one-sided formulas that can evaluate to logical
#' vectors.
#' The input is represented in these conditional statements as "`.`".
#' Valid formulas take the form `~ . < 0`.
#' See examples.
#'
#' @param x Vector to modify
#' @param ... Values to replace with `NA`, specified as either:
#' \itemize{
#'     \item An object, vector of objects, or list of objects
#'     \item A one-sided formula (see section "Formulas")
#' }
#' @return A modified version of `x` with selected values replaced with
#' `NA`.
#'
#' @seealso [dplyr::na_if()] to replace a single value with `NA`.
#'
#'   [dplyr::coalesce()] to replace missing values with a specified value.
#'
#'   [tidyr::replace_na()] to replace `NA` with a value.
#'
#'   [dplyr::recode()] and [dplyr::case_when()] to more generally replace
#'   values.
#'
#' @export
#' @examples
#' -1:10
#' # We can replace -1...
#' # ... explicitly
#' na_if_in(-1:10, -1)
#' # ... by specifying values to keep
#' na_if_not(-1:10, 0:10)
#' # ... using a formula
#' na_if_in(-1:10, ~ . < 0)
#' # ... or using a function
#' na_if_in(-1:10, min)
#'
#' messy_string <- c("abc", "", "def", "NA", "ghi", 42, "jkl", "NULL", "mno")
#' # We can replace unwanted values...
#' # ... one at a time
#' na_if_in(messy_string, "")
#' # ... or all at once
#' na_if_in(messy_string, "", "NA", "NULL", 1:100)
#' na_if_in(messy_string, c("", "NA", "NULL", 1:100))
#' na_if_in(messy_string, list("", "NA", "NULL", 1:100))
#' # ... or using a clever formula
#' grepl("[a-z]{3,}", messy_string)
#' na_if_not(messy_string, ~ grepl("[a-z]{3,}", .))
#'
#' # na_if_in() is particularly useful inside dplyr::mutate
#' library(dplyr)
#' faux_census %>%
#'   mutate(
#'     state = na_if_in(state, "Canada"),
#'     age   = na_if_in(age, ~ . < 18, ~ . > 120)
#'   )
#'
#' # We get a message if our values to replace don't exist
#' na_if_in(-1:10, 11)
#' # And a warning if we use an invalid input...
#' # ... like a two-sided formula
#' na_if_in(-1:10, x ~ . < 0)
#' # ... NULL
#' na_if_in(-1:10, NULL)
#' # ... or nothing at all
#' na_if_in(-1:10)
#'
#' # This function handles vector values differently than dplyr,
#' # and returns a different result with vector replacement values:
#' na_if_in(1:5, 5:1)
#' dplyr::na_if(1:5, 5:1)

na_if_in <- function(x, ...) {
  faux_na_if(x, ...)
}

#' @rdname na_if_in
#' @details # Lifecycle
#'
#' \lifecycle{deprecated} `na_if()` has been deprecated in favor of
#' `na_if_in()` to avoid masking [dplyr::na_if()].
#'
#' @export

na_if <- function(x, ...) {
  if (length(sys.call()) > 3 || !is.atomic(...)) {
    lifecycle::deprecate_warn("0.6.0", "fauxnaif::na_if()", "na_if_in()")
  }

  faux_na_if(x, ...)
}

#' @rdname na_if_in
#' @export

na_if_not <- function(x, ...) {
  faux_na_if(x, ..., not = TRUE)
}
